File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInsidePresentationControllersAdaptivityandCustomAnimatorObjects/LookInside/AAPLOverlayViewController.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  AAPLOverlayViewController implementation.
  
 */

#import "AAPLOverlayViewController.h"

@interface AAPLOverlayViewController ()
@property (nonatomic, strong) CIContext *context;
@property (nonatomic, strong) CIImage *baseCIImage;
@property (nonatomic, strong) CIFilter *colorControlsFilter;
@property (nonatomic, strong) CIFilter *hueAdjustFilter;
@end

@implementation AAPLOverlayViewController
{
    dispatch_queue_t processingQueue;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if(self)
    {
        [self setModalPresentationStyle:UIModalPresentationCustom];
        [self setup];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    foregroundContentView = [[UIVisualEffectView alloc] initWithEffect:[UIVibrancyEffect effectForBlurEffect:blurEffect]];
    backgroundView = [[UIVisualEffectView alloc] initWithEffect:blurEffect];

    [self configureViews];
}

- (void)configureCIObjects {
    if (!self.context) {
        self.context = [CIContext contextWithOptions:nil];
    }
    
    self.baseCIImage = [CIImage imageWithCGImage:[[[self photoView] image] CGImage]];
}

- (void)setPhotoView:(AAPLPhotoCollectionViewCell *)photoView {
    if (currentPhotoView != photoView) {
        currentPhotoView = photoView;
        
        [self configureCIObjects];
    }
}

- (void)sliderChanged:(id)sender
{
    CGFloat hue = [hueSlider value];
    CGFloat saturation = [saturationSlider value];
    CGFloat brightness = [brightnessSlider value];

    // Update labels

    [hueLabel setText:[NSString stringWithFormat:NSLocalizedString(@"Hue: %f", @"Hue label format."), hue]];
    [saturationLabel setText:[NSString stringWithFormat:NSLocalizedString(@"Saturation: %f", @"Saturation label format."), saturation]];
    [brightnessLabel setText:[NSString stringWithFormat:NSLocalizedString(@"Brightness: %f", @"Brightness label format."), brightness]];

    // Apply effects to image
    
    dispatch_async(processingQueue, ^{
        if (!self.colorControlsFilter) {
            self.colorControlsFilter = [CIFilter filterWithName:@"CIColorControls"];
        }
        [self.colorControlsFilter setValue:self.baseCIImage forKey:kCIInputImageKey];
        [self.colorControlsFilter setValue:@(saturation) forKey:@"inputSaturation"];
        [self.colorControlsFilter setValue:@(brightness) forKey:@"inputBrightness"];
        
        CIImage *coreImageOutputImage = [self.colorControlsFilter valueForKey:kCIOutputImageKey];
        
        if (!self.hueAdjustFilter) {
            self.hueAdjustFilter = [CIFilter filterWithName:@"CIHueAdjust"];
        }
        [self.hueAdjustFilter setValue:coreImageOutputImage forKey:kCIInputImageKey];
        [self.hueAdjustFilter setValue:@(hue) forKey:@"inputAngle"];
        
        coreImageOutputImage = [self.hueAdjustFilter valueForKey:kCIOutputImageKey];
        
        CGImageRef cgImage = [self.context createCGImage:coreImageOutputImage fromRect:CGRectMake(0,0,[[[self photoView] image] size].width, [[[self photoView] image] size].height)];
        UIImage *image = [UIImage imageWithCGImage:cgImage];
        CGImageRelease(cgImage);
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [imageView setImage:image];
        });
    });
    
}

- (void)savePushed:(id)sender
{
    [[self photoView] setImage:[imageView image]];
    [[self presentingViewController] dismissViewControllerAnimated:YES completion:NULL];
}

- (UISlider *)configuredOverlaySlider
{
    UISlider *slider = [[UISlider alloc] init];
    [slider setTranslatesAutoresizingMaskIntoConstraints:NO];
    [slider addTarget:self action:@selector(sliderChanged:) forControlEvents:UIControlEventValueChanged];
    [slider setContinuous:NO];
    return slider;
}

- (void)setup
{
    imageView = [[UIImageView alloc] init];
    [imageView setContentMode:UIViewContentModeScaleAspectFit];
    [imageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleExtraLight];
    
    processingQueue = dispatch_queue_create("image processing queue", DISPATCH_QUEUE_SERIAL);
}

- (void)configureViews
{
    [imageView setImage:[[self photoView] image]];
    [[self view] setBackgroundColor:[UIColor clearColor]];

    [backgroundView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [foregroundContentView setTranslatesAutoresizingMaskIntoConstraints:NO];
    hueLabel = [[AAPLOverlayVibrantLabel alloc] init];
    hueSlider = [self configuredOverlaySlider];
    [hueSlider setMaximumValue:10.0];

    saturationLabel = [[AAPLOverlayVibrantLabel alloc] init];
    saturationSlider = [self configuredOverlaySlider];
    [saturationSlider setValue:1.0];
    [saturationSlider setMaximumValue:2.0];
    
    brightnessLabel = [[AAPLOverlayVibrantLabel alloc] init];
    brightnessSlider = [self configuredOverlaySlider];
    [brightnessSlider setMinimumValue:-0.5];
    [brightnessSlider setMaximumValue:0.5];

    saveButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [saveButton setTranslatesAutoresizingMaskIntoConstraints:NO];
    [saveButton setTitle:NSLocalizedString(@"Save", @"Save button title.") forState:UIControlStateNormal];
    [[saveButton titleLabel] setFont:[UIFont systemFontOfSize:32.0]];
    [saveButton addTarget:self action:@selector(savePushed:) forControlEvents:UIControlEventTouchUpInside];
    
    [[self view] addSubview:backgroundView];
    [[self view] addSubview:foregroundContentView];

    [[foregroundContentView contentView] addSubview:hueLabel];
    [[foregroundContentView contentView] addSubview:hueSlider];

    [[foregroundContentView contentView] addSubview:saturationLabel];
    [[foregroundContentView contentView] addSubview:saturationSlider];

    [[foregroundContentView contentView] addSubview:brightnessLabel];
    [[foregroundContentView contentView] addSubview:brightnessSlider];

    [[foregroundContentView contentView] addSubview:saveButton];

    [[self view] addSubview:imageView];
    
    // add constraints
    NSDictionary *views = NSDictionaryOfVariableBindings(backgroundView, foregroundContentView, hueLabel, hueSlider, saturationLabel, saturationSlider, brightnessLabel, brightnessSlider, saveButton, imageView);
    
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[backgroundView]|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[backgroundView]|" options:0 metrics:nil views:views]];
    
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[foregroundContentView]|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[foregroundContentView]|" options:0 metrics:nil views:views]];
    
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[hueLabel]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[hueSlider]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[saturationLabel]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[saturationSlider]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[brightnessLabel]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[brightnessSlider]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[saveButton]-|" options:0 metrics:nil views:views]];
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[imageView]|" options:0 metrics:nil views:views]];
    
    [[self view] addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(>=30)-[hueLabel]-[hueSlider]-[saturationLabel]-[saturationSlider]-[brightnessLabel]-[brightnessSlider]-[saveButton]-(>=10)-[imageView]|" options:0 metrics:nil views:views]];
    
    [self sliderChanged:nil];
}

@end

@implementation AAPLOverlayVibrantLabel

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if(self)
    {
        [self setTranslatesAutoresizingMaskIntoConstraints:NO];
    }
    
    return self;
}

- (void)tintColorDidChange
{
    [self setTextColor:[self tintColor]];
}

@end

File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInside_Swift/LookInside_Swift/AAPLOverlayViewController.swift
//
//  AAPLOverlayViewController.swift
//  LookInside_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/18/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

class AAPLOverlayViewController: UIViewController {

	var backgroundView: UIVisualEffectView!
	var foregroundContentView: UIVisualEffectView!
	
	var blurEffect: UIBlurEffect!
	var imageView: UIImageView!
	
	var hueLabel: AAPLOverlayVibrantLabel!
	var hueSlider: UISlider!
	
	var saturationLabel: AAPLOverlayVibrantLabel!
	var saturationSlider: UISlider!

	var brightnessLabel: AAPLOverlayVibrantLabel!
	var brightnessSlider: UISlider!

	var saveButton: UIButton!
	var currentPhotoView: AAPLPhotoCollectionViewCell!
	
	var photoView: AAPLPhotoCollectionViewCell! {
		didSet {
			if self.currentPhotoView != photoView {
				self.currentPhotoView = photoView
			}
			self.configureCIObjects()
		}
	}

	var context: CIContext!
	var baseCIImage: CIImage!
	var colorControlsFilter: CIFilter!
	var hueAdjustFilter: CIFilter!

	var processingQueue: DispatchQueue!
	
	override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
		super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
		
		self.modalPresentationStyle = .custom
		self.setup()
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
    override func viewDidLoad() {
        super.viewDidLoad()
		
		self.foregroundContentView = UIVisualEffectView(effect: UIVibrancyEffect(blurEffect: self.blurEffect))
		self.backgroundView = UIVisualEffectView(effect: self.blurEffect)
		self.configureViews()
    }
	
	func configureCIObjects() {
		if self.context == nil {
			self.context = CIContext(options: nil)
		}
		
		self.baseCIImage = CIImage(cgImage: self.photoView.image!.cgImage!)
	}
	
	@objc func sliderChanged(_ sender: Any?) {
		
		let hue: CGFloat = CGFloat(self.hueSlider.value)
		let saturation: CGFloat = CGFloat(self.saturationSlider.value)
		let brightness: CGFloat = CGFloat(self.brightnessSlider.value)
		
		// Update labels
		
		self.hueLabel.text = NSLocalizedString("Hue: \(hue)", comment: "Hue label format.")
		self.saturationLabel.text = NSLocalizedString("Saturation: \(saturation)", comment: "Saturation label format.")
		self.brightnessLabel.text = NSLocalizedString("Brightness: \(brightness)", comment: "Brightness label format.")

		// Apply effects to image
		let imageSize = self.photoView.image!.size
		self.processingQueue.async {
			if self.colorControlsFilter == nil {
				self.colorControlsFilter = CIFilter(name: "CIColorControls")
			}
			
			self.colorControlsFilter.setValue(self.baseCIImage, forKey: kCIInputImageKey)
			self.colorControlsFilter.setValue(saturation, forKey: "inputSaturation")
			self.colorControlsFilter.setValue(brightness, forKey: "inputBrightness")

			var coreImageOutputImage = self.colorControlsFilter.outputImage
			
			if self.hueAdjustFilter == nil {
				self.hueAdjustFilter = CIFilter(name: "CIHueAdjust")
			}
			self.hueAdjustFilter.setValue(coreImageOutputImage, forKey: kCIInputImageKey)
			self.hueAdjustFilter.setValue(hue, forKey: "inputAngle")
			
			coreImageOutputImage = self.hueAdjustFilter.outputImage
			
			let cgImage = self.context.createCGImage(coreImageOutputImage!, from: CGRect(x: 0, y: 0, width: imageSize.width, height: imageSize.height))
			let image = UIImage(cgImage: cgImage!)
			
			DispatchQueue.main.async {
				self.imageView.image = image
			}
		}
	}
	
	@objc func savePushed(_ sender: Any) {
		self.photoView.image = self.imageView.image
		self.presentingViewController?.dismiss(animated: true, completion: nil)
	}
	
	func configuredOverlaySlider() -> UISlider {
		let slider = UISlider()
		slider.translatesAutoresizingMaskIntoConstraints = false
		slider.addTarget(self, action: #selector(sliderChanged(_:)), for: .valueChanged)
		slider.isContinuous = false
		return slider
	}

	func setup() {
		self.imageView = UIImageView()
		self.imageView.contentMode = .scaleAspectFit
		self.imageView.translatesAutoresizingMaskIntoConstraints = false
		self.blurEffect = UIBlurEffect(style: .extraLight)
		
		self.processingQueue = DispatchQueue(label: "image processing queue")
	}
	
	func configureViews() {
		self.imageView.image = self.photoView.image
		self.view.backgroundColor = .clear
		
		self.backgroundView.translatesAutoresizingMaskIntoConstraints = false
		self.foregroundContentView.translatesAutoresizingMaskIntoConstraints = false
		
		self.hueLabel = AAPLOverlayVibrantLabel()
		self.hueSlider = self.configuredOverlaySlider()
		self.hueSlider.maximumValue = 10.0
		
		self.saturationLabel = AAPLOverlayVibrantLabel()
		self.saturationSlider = self.configuredOverlaySlider()
		self.saturationSlider.value = 1.0
		self.saturationSlider.maximumValue = 2.0

		self.brightnessLabel = AAPLOverlayVibrantLabel()
		self.brightnessSlider = self.configuredOverlaySlider()
		self.brightnessSlider.minimumValue = -0.5
		self.brightnessSlider.maximumValue = 0.5
		
		self.saveButton = UIButton(type: .system)
		self.saveButton.translatesAutoresizingMaskIntoConstraints = false
		self.saveButton.setTitle(NSLocalizedString("Save", comment: "Save button title."), for: .normal)
		self.saveButton.titleLabel?.font = UIFont.systemFont(ofSize: 32.0)
		self.saveButton.addTarget(self, action: #selector(savePushed(_:)), for: .touchUpInside)
		
		self.view.addSubview(self.backgroundView)
		self.view.addSubview(self.foregroundContentView)

		self.foregroundContentView.contentView.addSubview(self.hueLabel)
		self.foregroundContentView.contentView.addSubview(self.hueSlider)

		self.foregroundContentView.contentView.addSubview(self.saturationLabel)
		self.foregroundContentView.contentView.addSubview(self.saturationSlider)

		self.foregroundContentView.contentView.addSubview(self.brightnessLabel)
		self.foregroundContentView.contentView.addSubview(self.brightnessSlider)

		self.foregroundContentView.contentView.addSubview(self.saveButton)

		self.view.addSubview(self.imageView)
		
		// add constraints
		let views = ["backgroundView" : backgroundView, "foregroundContentView" : foregroundContentView, "hueLabel" : hueLabel, "hueSlider" : hueSlider, "saturationLabel" : saturationLabel, "saturationSlider" : saturationSlider, "brightnessLabel" : brightnessLabel, "brightnessSlider" : brightnessSlider, "saveButton" : saveButton, "imageView" : imageView] as [String : Any]
		
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[backgroundView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[backgroundView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))

		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[foregroundContentView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[foregroundContentView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))

		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[hueLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[hueSlider]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[saturationLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[saturationSlider]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[brightnessLabel]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[brightnessSlider]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-[saveButton]-|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[imageView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))
		
		self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-(>=30)-[hueLabel]-[hueSlider]-[saturationLabel]-[saturationSlider]-[brightnessLabel]-[brightnessSlider]-[saveButton]-(>=10)-[imageView]|", options: NSLayoutFormatOptions(), metrics: nil, views: views))

		self.sliderChanged(nil)
	}
	
}

class AAPLOverlayVibrantLabel: UILabel {
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		self.translatesAutoresizingMaskIntoConstraints = false
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	override func tintColorDidChange() {
		self.textColor = self.tintColor
	}
}
